apps/shared/avifpng.c

```diff
@@ -17,6 +20,119 @@ typedef png_bytep png_iccp_datap;
+    for (; (actualLength < expectedLength) && (*src != '\0'); ++actualLength) {
+        while (*src == '\n') {
+            ++src;
+        }
+        if (!isxdigit(src[0]) || !isxdigit(src[1])) {
+            avifRWDataFree(bytes);
+            fprintf(stderr, "Exif extraction failed: invalid token at " AVIF_FMT_ZU "\n", actualLength);
+            return AVIF_FALSE;
+        }
+        const char twoHexDigits[] = { src[0], src[1], '\0' };
+        bytes->data[actualLength] = (uint8_t)strtol(twoHexDigits, NULL, 16);
+        src += 2;
+    }
+
+    if (actualLength != expectedLength) {
+        avifRWDataFree(bytes);
+        fprintf(stderr, "Exif extraction failed: expected " AVIF_FMT_ZU " tokens but got " AVIF_FMT_ZU "\n", expectedLength, actualLength);
+        return AVIF_FALSE;
+    }
+    return AVIF_TRUE;
+}
+
+// Parses the raw profile string of profileLen characters and extracts the payload.
+static avifBool avifCopyRawProfile(const char * profile, size_t profileLength, avifRWData * payload)
+{
+    // ImageMagick formats 'raw profiles' as "\n<name>\n<length>(%8lu)\n<hex payload>\n".
+    if (!profile || (profileLength < 3) || (profile[0] != '\n')) {
+        fprintf(stderr, "Exif extraction failed: truncated or malformed raw profile\n");
+        return AVIF_FALSE;
+    }
+
+    const char * lengthStart = NULL;
+    for (size_t i = 1; i < profileLength; ++i) {
+        if (profile[i] == '\0') {
+            fprintf(stderr, "Exif extraction failed: malformed raw profile, unexpected null-terminating character at " AVIF_FMT_ZU "\n", i);
+            return AVIF_FALSE;
+        } else if (profile[i] == '\n') {
+            if (!lengthStart) {
+                // Skip the name and store the beginning of the string containing the length of the payload.
+                lengthStart = &profile[i + 1];
+            } else {
+                const char * hexPayloadStart = &profile[i + 1];
+                const size_t hexPayloadMaxLength = profileLength - (i + 1);
+                // Parse the length, now that we are sure that it is surrounded by '\n' within the profileLength characters.
+                char * lengthEnd;
+                const long expectedLength = strtol(lengthStart, &lengthEnd, 10);
+                if (lengthEnd != &profile[i]) {
+                    fprintf(stderr, "Exif extraction failed: malformed raw profile, expected '\\n' but got '\\x%.2X'\n", *lengthEnd);
+                    return AVIF_FALSE;
+                }
+                // No need to check for errno. Just make sure expectedLength is not LONG_MIN and not LONG_MAX.
+                if ((expectedLength <= 0) || (expectedLength == LONG_MAX) || ((unsigned long)expectedLength > hexPayloadMaxLength)) {
+                    fprintf(stderr, "Exif extraction failed: invalid length %ld\n", expectedLength);
+                    return AVIF_FALSE;
+                }
+                return avifHexStringToBytes(hexPayloadStart, (size_t)expectedLength, payload);
```

Comment:
Nit: This is just a comment.

`expectedLength` may be much smaller than `profileLength`, so that there are extra hexadecimal digits in the payload after we have `expectedLength` converted bytes. It seems that this kind of payload is also malformed.

I am just pointing this out for your reference, because I don't know if you aim to reject all malformed payloads.